akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala

```diff
@@ -0,0 +1,247 @@
+
+  sealed trait CircuitBreakerState
+  case object Open extends CircuitBreakerState
+  case object Closed extends CircuitBreakerState
+  case object HalfOpen extends CircuitBreakerState
+
+  final case class CircuitBreakerStateData(failureCount: Int = 0, firstHalfOpenMessageSent: Boolean = false)
+
+  final case class CircuitBreakerActorBuilder(
+    maxFailures: Int, callTimeout: Timeout, resetTimeout: Timeout,
+    circuitEventListener: Option[ActorRef] = None,
+    failureDetector: Any ⇒ Boolean = { _ ⇒ false },
+    openCircuitFailureConverter: CircuitOpenFailure ⇒ Any = identity) {
+
+    /**
+     * Creates the props for a [[akka.contrib.circuitbreaker.CircuitBreakerActor]] proxying the given target
+     *
+     * @param target the target actor ref
+     */
+    def propsForTarget(target: ActorRef) = CircuitBreakerActor.props(target, maxFailures, callTimeout, resetTimeout, circuitEventListener, failureDetector, openCircuitFailureConverter)
+
+  }
+
+  class OpenCircuitException extends Exception("Circuit Open so unable to complete operation")
+
+  /**
+   * Extends [[scala.concurrent.Future]] with the method failForOpenCircuitWith to handle
+   *   [[akka.contrib.circuitbreaker.CircuitBreakerActor.CircuitOpenFailure]] failure responses throwing
+   *   an exception built with the given exception builder
+   */
+  implicit class CircuitBreakerAwareFuture(val future: Future[Any]) extends AnyVal {
+    def failForOpenCircuit(implicit executionContext: ExecutionContext): Future[Any] = failForOpenCircuitWith(new OpenCircuitException)
+
+    def failForOpenCircuitWith(throwing: ⇒ Throwable)(implicit executionContext: ExecutionContext): Future[Any] = {
+      future.flatMap {
+        _ match {
+          case CircuitOpenFailure(_) ⇒ Future.failed(throwing)
+          case result                ⇒ Future.successful(result)
+        }
+      }
+    }
+
+  }
+
+}
+
+object CircuitBreakerInternalEvents {
+  sealed trait CircuitBreakerInternalEvent
+  case object CallFailed extends CircuitBreakerInternalEvent
+  case object CallSucceeded extends CircuitBreakerInternalEvent
+}
+
+import CircuitBreakerActor._
+import CircuitBreakerInternalEvents._
+
+class CircuitBreakerActor(
+  target: ActorRef,
+  maxFailures: Int,
+  callTimeout: Timeout,
+  resetTimeout: Timeout,
+  circuitEventListener: Option[ActorRef],
+  failureDetector: Any ⇒ Boolean,
+  failureMap: CircuitOpenFailure ⇒ Any) extends Actor with ActorLogging with FSM[CircuitBreakerState, CircuitBreakerStateData] {
+
+  startWith(Closed, CircuitBreakerStateData(failureCount = 0))
+
+  def callSucceededHandling: StateFunction = {
+    case Event(CallSucceeded, state) ⇒
+      log.debug("Received call succeeded notification in state {} resetting counter", state)
+      goto(Closed) using CircuitBreakerStateData(failureCount = 0, firstHalfOpenMessageSent = false)
+  }
+
+  when(Closed) {
+    callSucceededHandling orElse {
```

Comment:
`CallSucceeded` seems unlikely to arrive when in `Closed` state as I read it. Shall we re-order here, such that the most frequent message is first (the partial function, then `orElse callSucceededHandling`)?